BrowserWidget Selection

BryonLewis opened the pull request:
Provides the ability to pass a Resource’s Model to the **BrowserWidget** and for that item to be selected and scrolled to automatically when the **BrowserWidget** loads.  It will also maintain scroll position through the addition of items and images to the **ItemListView** (typically through plugins adding additional content).  
NOTE:  This doesn’t currently have support for the FolderListWidget, but it can be added after this style and method is approved for the ItemListWidget.  Then these methods could be abstracted out of the ItemListWidget and shared by both it and the FolderListWidget

BrowserWidget has two additional options: **defaultSelectedResource** (Model), **highlightItem** (string).
**defaultSelectedResource** will launch the BrowserWidget with the RootSelectorWidget populated to the baseParentId of the defaultSelectedResource and the HierarchyWidget will be on the parent folder of the defaultSelectedResource.
**highlightItem** will make the defaultSelectedResource item, or any selectedItem become styled with the selected class in the HierarchyWidget as well as scroll to center the selection in the list.  It will maintain the centered position even when other elements are added to the list through plugins, like images and additional UI elements.

A majority of the changes are within the RootSelectorWidget and the ItemListWidget.  The RootSelectorWidget will now navigate to the selected RootParent item either by using a model which contains baseParentId or by giving it a baseParentId string directly.
The ItemListWidget now creates a DOM mutation observer which watches for DOM additions after it is initially rendered.  It then will adjust the scroll position to the selected item when DOM operations are complete.  It will also monitor for the additional of and IMG tags and create a function that will adjust the scroll position once the IMG is loaded.  Finally it monitors for user scrolling and will prevent any automatic scrolling if the user initiates a scroll before the DOM and images are fully loaded.

Performance:
- Set the RootSelectorWidget by Id requires going through all the groups and issuing a getByID call to see if the parent ID is in one of them.  Can be an issue with a larger number of groups although get(id) is utilizing key structures so it isn't that bad.  With the standard default groups this could be cut down because the resourceName is typically the index key to the master ‘groups’ object, but that isn’t the case for custom groups.
- Mutation monitoring and scrolling - I tried to reduce the number of times the scrolling occurs at least with respect to adding standard elements.  The initial call to requestIdleCallback should do an initial position after all simple DOM elements are added.  Then the series of other calls are when the images are properly loaded, it does check to make sure the scroll position is different before attempting to move it.  Also there can be penalty for initiating this many requests to the DOM through jquery for positional information as well as manually setting the scroll position.

manthey commented on girder/web_client/src/views/widgets/ItemListWidget.js:
```diff
@@ -140,6 +147,110 @@ var ItemListWidget = View.extend({
+                this.tempScrollPos = scrollPos;
+            }
+            // Call a custom scroll event to prevent thinking the user initiated it
+            var e = new CustomEvent('scroll', { detail: 9999 });
```
Instead of using `9999`, I think this would be more obvious if it were an explicit string, like `browser_widget_scroll_event`, unless there is some reason this needs to be a number.

manthey commented on girder/web_client/src/views/widgets/ItemListWidget.js:
```diff
@@ -140,6 +147,110 @@ var ItemListWidget = View.extend({
+                            this.observer.disconnect();
+                            this.observer = null;
+                            this.tempScrollPos = undefined;
+                        }
```
I think it would be clearer if this and the equivalent block below were in their own function rather than repeated.

manthey commented on girder/web_client/src/views/widgets/ItemListWidget.js:
```diff
@@ -140,6 +147,110 @@ var ItemListWidget = View.extend({
+                                // add the event listener to the onload of each image
+                                imgs[i].addEventListener('load', onLoadImage);
+                            }
+                        }
```
As long as we are using jQuery, this block could be changed to `$(node).on('load', 'img', onLoadImage)`.  If has the virtue that in new img tags are injected by plugins, they would be watched as well.

BryonLewis replied:
I've tried to add this but it doesn't seem to call the handler properly.  I could be wrong but for jQuery event delegation to work properly the event has to bubble?   [Jquery .on Documentation](https://api.jquery.com/on/), in the notes at the bottom it specifically talks about how load events don't bubble up.

> In all browsers, the load, scroll, and error events (e.g., on an <img> element) do not bubble. In Internet Explorer 8 and lower, the paste and reset events do not bubble. Such events are not supported for use with delegation, but they can be used when the event handler is directly attached to the element generating the event.
> 
Through the mutation observer all additional img elements should be caught as long as the user doesn't move the scroll area before they are added.  I don't kill the observer until there is user interaction (any scroll not initiated by the program).

manthey replied:
You're right.  `$('img', node),on('load', onLoadImage)` would work, though that is probably doing exactly what your code is doing anyway.

manthey commented on girder/web_client/src/views/widgets/ItemListWidget.js:
```diff
@@ -140,6 +147,106 @@ var ItemListWidget = View.extend({
+            // Backup function to kill observer if user moves scrollwheel, or touchpad equivalent
+            widgetcontainer.bind('wheel.selectionobserver', (evt) => {
+                unbindDisconnect();
+            });
```
This can now be `widgetcontainer.bind('wheel.selectionobserver', unbindDisconnect);`

BryonLewis commented on girder/web_client/src/views/widgets/BrowserWidget.js:
```diff
@@ -75,17 +82,24 @@ var BrowserWidget = View.extend({
             this._renderHierarchyView();
         });
+        if (this.defaultSelectedResource && this.defaultSelectedResource.parent) {
+            this.root = this.defaultSelectedResource.parent;
```
Just some planning for going about doing this:

So the original default was to have no root set and by default it would just present something that is not selected.  I think it is best to calculate the parent if we can to reduce friction from using the 'selectedDefaultResource' parameter with the browser widget.

Some Background Information for myself:
Item - Can get a `folderId` and fetch to get the parent model
Folder - parent is either a folder/collection/user: `parentCollection` ('folder'|'collection'|'user') and `parentId`, meaning I have to conditionally fetch each different type of model and set the root to that instead.  
**Question**: Looking through hHerarchyWidget it references the root (parentModel) 92 times, so I'm assuming through all the functions it needs an explicit version of the specific model type (Item,Collection,User,Folder) and not some generic parent?

My proposed process for this:
1. Check to see if a `root` is already assigned, if so default to that.
2. If a `defaultSelectedResource` is assigned and already has an existing `parent`, use that as the new root
3. If no `parent` exists determine if it has a `folderId` associated with the defaultSelectedResource, followed by creating a FolderModel and fetching the data, on fetch set the root to the FolderModel
4. If no folderId exists check for a `parentCollection` and `parentId`, depending on those values either create and fetch a `UserModel`, `CollectionModel`, or `FolderModel` then assign root to that
5. finally if none of the above exists we default to the undefined root

Additional Self Notes:
Now there are ways I should be able to generalize the Model retrieval and collection, they should all have `.on('g:fetched')` and on.('g:error').  Possibly an object with the `parentCollection` types as a key to the different Model constructors.
Also just default to using the collectionParent type folder, for if the folderId exists instead of duplicating code for retrieval of the model.

manthey replied:
folders all have baseParentId and baseParentType that will be a collection or user, so the `rootSelectorSettings` can be `{selectbyResource: <folder or item's folder>.get('baseParentId')}`.

manthey replied:
Because of the virtual folders plugin, it is useful to be able to provide the intended parent.  Having it default to the item's `folderId` would be handy use cases which aren't virtual-folder aware, though.  Since `item.parent` is used for access control, I'm not sure of the repercussions of overloading it to do this.  It might be better to have another option for this.

BryonLewis replied:
> folders all have baseParentId and baseParentType that will be a collection or user, so the `rootSelectorSettings` can be `{selectbyResource: <folder or item's folder>.get('baseParentId')}`.

Yes for the rootSelectorWidget I could just use that, but I do need the parent of the current model for the HierarchyWidget. In the master branch the hierarachyWidget within browserWidget uses the root as the parentModel, for displaying the folder and the breadcrumb trail of the current folder.
![image](https://user-images.githubusercontent.com/61746913/77076793-b3dae980-69ca-11ea-8a89-173911428feb.png)

Before I don't think this was an issue because the root was a collection/user.  Now it needs to be a collection or folder to open to the proper location initially.  I can change this in the browserWidget to use my own value instead of `this.root` when a selectedDefaultResource Model is available.

manthey replied:
Agreed.  The `baseParentId` only helps the `rootSelectorSettings`, not the HierarchyWidget's parentModel.

BryonLewis commented:
- I've added in support for opening default folders with the BrowserWidget now.  You can specify a folder Model as the defaultSelectedResource and it will directly open up to that folder with it already selected.
- The root Model is now calculated directly through the `folderId`, `parentCollection` and `parentId`.  Of note is that for most items the `parentCollection` will typically be a folder and if folders are the selected type that it will be it's own `root`.  That's the way the folder selection scheme works, you open the folder you want selected not the parent folder.
- I've updated the `highlightItem` field and passed it down to the proper UI items to maintain the non-selected default nature of the browser widget.  Before this change anything with a specified defaultSelectedResource would have highlighting for the item enabled by default.  This allows the highlighting and scrolling to be disabled by default and enabled by setting `highlightItem` to true
- You can also manually specify the `root` item to be a virtual folder for selecting the item and it will open to the item with the parent folder being the virtual folder.  This can only be done through manually setting the root.
- Finally I updated the testing to fully include the new browser widget settings for folders, items and highlighting.

manthey commented:
I'll argue that this really does belong in master and not as a plugin.

Having the ability to show what is already selected is highly desirable.  We have a number of places where we store a resource location (the location of a folder as part of a plugin setting, the location of an item as part of a UI).  With this PR, these controls can start at the selected item, which is vastly more expected than starting at the root selector without any hint of where the previously selected folder/item was located.

Since this is a desired feature, if it were in a plugin, it seems like many consumers of the BrowserWidget would choose to have that plugin enabled.  Assuming that, the plugin would add more maintenance burden than the same code in core.

More review is good if we can do it in a timely fashion.

manthey commented on girder/web_client/src/views/widgets/BrowserWidget.js:
```diff
@@ -4,6 +4,9 @@ import _ from 'underscore';
 import View from '@girder/core/views/View';
+import CollectionModel from '@girder/core/models/CollectionModel';
+import FolderModel from '@girder/core/models/FolderModel';
+import  UserModel from '@girder/core/models/UserModel';
```
Extra space after `import`.  You'd think the linter would complain, but it didn't.

manthey commented on girder/web_client/src/views/widgets/BrowserWidget.js:
```diff
@@ -238,6 +268,46 @@ var BrowserWidget = View.extend({
+            // Case for providing a folder as the defaultSelectedResource but want the user to select an item
+            // folder parent is either 'user' | 'folder' | 'collection', most likely folder though
+            modelType = this.defaultSelectedResource.get('parentCollection');
+            modelId = this.defaultSelectedResource.get('parentId');
```
Can you add to the tests to cover this condition, too?

BryonLewis replied:
New tests added should cover this specific case as well.

manthey commented on girder/web_client/src/views/widgets/ItemListWidget.js:
```diff
@@ -140,6 +151,104 @@ var ItemListWidget = View.extend({
+                    this.observer = null;
+                    this.tempScrollPos = undefined;
+                }
+            };
```
Can we think of a way to include this in the tests?  It may be far too much work to be worth it.

BryonLewis replied:
As an extra precaution I'm adding a click observer to unbindDisconnect as well for the case in which a user tries to select a different item while images are still loading.  The problem with the standard scrolling is creating that many items in Jasmine usually takes too long (or I would have to restructure how I create items) and I don't think I can trigger an actual scrollevent in a non-scrolling element.  I can easily trigger a click event like I am clicking or interacting and that can trigger the unbind.  That is something I have a test for.

manthey replied:
Understood.

You can create items from the rest endpoint directly (calling `girder.rest.restRequest` with a `POST item`, rather than mimicking user behavior), though that may also be slower than desired.

zachmullen commented on girder/web_client/src/views/widgets/BrowserWidget.js:
```diff
@@ -238,6 +268,46 @@ var BrowserWidget = View.extend({
+
+    /**
+     * If we have a defaultSelectedResource we need the root item for the hiearachyWidget
+     * This will calcualte what the root item should be if one hasn't been supplied
```
calculate*

zachmullen commented on girder/web_client/src/views/widgets/ItemListWidget.js:
```diff
@@ -140,6 +151,108 @@ var ItemListWidget = View.extend({
+                this.tempScrollPos = scrollPos;
+            }
+            // Call a custom scroll event to prevent thinking the user initiated it
+            var e = new CustomEvent('scroll', { detail: 'selected_item_scroll' });
```
Not sure what our linter settings are, but it looks like it's not disallowing `var`. Could you change this to `const`?

zachmullen commented on girder/web_client/src/views/widgets/ItemListWidget.js:
```diff
@@ -140,6 +151,108 @@ var ItemListWidget = View.extend({
+
+    centerSelected: function () {
+        const widgetcontainer = $('.g-hierarchy-widget-container');
+        const selected = $('li.g-item-list-entry.g-selected');
```
In general, we should use `this.$` instead of `$` whenever we're referring to elements underneath the current component, but it looks like the first one of these may be referring to a parent component. In that case, we should still be looking relative to ourselves rather than globally, so we can use `this.$el.parents('.g-hierarchy-widget-container')`.

zachmullen commented on girder/web_client/src/views/widgets/ItemListWidget.js:
```diff
@@ -140,6 +151,108 @@ var ItemListWidget = View.extend({
+     */
+    scrollPositionObserver: function () {
+        // Set the default selected height for the selected item
+        const target = $('.g-item-list');
```
same here (relative vs. global selectors)

zachmullen commented on girder/web_client/src/views/widgets/ItemListWidget.js:
```diff
@@ -140,6 +151,108 @@ var ItemListWidget = View.extend({
+        const target = $('.g-item-list');
+        if (window.MutationObserver && target.length > 0) {
+            // Default items to monitor for the scroll position
+            const widgetcontainer = $('.g-hierarchy-widget-container');
```
same here (relative vs. global selectors)

zachmullen commented on girder/web_client/src/views/widgets/ItemListWidget.js:
```diff
@@ -140,6 +151,108 @@ var ItemListWidget = View.extend({
+
+                        // For any images added we wait until loaded and rescroll to the selected
+                        if (_.isFunction(node.getElementsByTagName)) {
+                            $('img', node).on('load', onLoadImage);
```
same here (`this.$('img')`)

zachmullen commented:
I haven't actually tested this, still deferring to @manthey on that. But my requests have been addressed.

zachmullen commented:
Not sure what kind of dark magic is needed for phantomjs... is this at least working in a real browser?

BryonLewis commented:
@zachmullen & @manthey  - sorry about the code change but there was some randomness in how phantomJS bound the scroll events that are used to maintain scrolling position.  I updated the test to try to use it the best it can but to skip it if the event is unbound (and give a small message).  I have not seen this issue occur in my testing of Chrome or Firefox with the scroll event handler.

zachmullen commented:
This is one of the downsides of using a no-longer-supported test environment. Thanks for pushing through the pain.

If this ends up failing nondeterministically it's probably best to just delete the test.
